source/generators/LambertianGenerator.cc

```diff
@@ -0,0 +1,150 @@
+using namespace CLHEP;
+
+REGISTER_CLASS(LED450LWcone, G4VPrimaryGenerator)
+
+
+LED450LWcone::LED450LWcone():
+G4VPrimaryGenerator(), msg_(0), particle_definition_(0),
+energy_min_(0.), energy_max_(0.), geom_(0), momentum_{},
+costheta_min_(-1.), costheta_max_(1.), phi_min_(0.), phi_max_(2.*pi)
+{
+  msg_ = new G4GenericMessenger(this, "/Generator/LED450LWcone/",
+    "Control commands of single-particle generator.");
+
+  msg_->DeclareMethod("particle", &LED450LWcone::SetParticleDefinition,
+    "Set particle to be generated.");
+
+  G4GenericMessenger::Command& min_energy =
+    msg_->DeclareProperty("min_energy", energy_min_,
+      "Set minimum kinetic energy of the particle.");
+  min_energy.SetUnitCategory("Energy");
+  min_energy.SetParameterName("min_energy", false);
+  min_energy.SetRange("min_energy>0.");
+
+  G4GenericMessenger::Command& max_energy =
+    msg_->DeclareProperty("max_energy", energy_max_,
+      "Set maximum kinetic energy of the particle");
+  max_energy.SetUnitCategory("Energy");
+  max_energy.SetParameterName("max_energy", false);
+  max_energy.SetRange("max_energy>0.");
+
+  msg_->DeclareProperty("region", region_,
+    "Set the region of the geometry where the vertex will be generated.");
+
+
+  msg_->DeclarePropertyWithUnit("momentum", "mm",  momentum_, "Set particle 3-momentum.");
+
+  msg_->DeclareProperty("min_costheta", costheta_min_,
+			"Set minimum cosTheta for the direction of the particle.");
+  msg_->DeclareProperty("max_costheta", costheta_max_,
+			"Set maximum cosTheta for the direction of the particle.");
+  msg_->DeclareProperty("min_phi", phi_min_,
+			"Set minimum phi for the direction of the particle.");
+  msg_->DeclareProperty("max_phi", phi_max_,
+			"Set maximum phi for the direction of the particle.");
+
+
+  DetectorConstruction* detconst = (DetectorConstruction*) G4RunManager::GetRunManager()->GetUserDetectorConstruction();
+  geom_ = detconst->GetGeometry();
+}
+
+
+
+LED450LWcone::~LED450LWcone()
+{
+  delete msg_;
+}
+
+
+
+void LED450LWcone::SetParticleDefinition(G4String particle_name)
+{
+  particle_definition_ =
+    G4ParticleTable::GetParticleTable()->FindParticle(particle_name);
+
+  if (!particle_definition_)
+    G4Exception("[LED450LW]", "SetParticleDefinition()",
+      FatalException, "User gave an unknown particle name.");
+}
+
+
+
+void LED450LWcone::GeneratePrimaryVertex(G4Event* event)
+{
+  // Generate uniform random energy in [E_min, E_max]
+  G4double kinetic_energy = nexus::UniformRandomInRange(energy_max_, energy_min_);
+
+  // Calculate cartesian components of momentum
+  G4double mass   = particle_definition_->GetPDGMass();
+  G4double energy = kinetic_energy + mass;
+  G4double pmod = std::sqrt(energy*energy - mass*mass);
+
+  G4ThreeVector dir = momentum_;
+
+  G4ThreeVector p_dir = G4LambertianRand(dir);
+
+  G4double cosine = -1*p_dir.dot(momentum_);
+
+  G4ThreeVector p;
+
+  while((costheta_min_ > cosine) || (cosine > costheta_max_)){
```

Comment:
From this `while` statement I deduce that the range of `theta` angles that the user must provide is about the angle with the "axis" of the Lambertian distribution and not, for instance, the global reference system. I think that this can originate errors (in other generators the range is on the reference system), therefore you should document it very clearly. Moreover, the `phi` range, also provided by the user, is not used in the code. If you want to leave the possibility of limiting `phi`, as well, you have to include it in the `while` statement.

Reply:
I've attached a picture to make the explanation clearer. 
Also, regarding the negative sign, I'm not sure how to make it more general. I've also deleted the feature of limiting `phi`. I don't find it necessary right  now and I find it difficult to include in a general way.

![angles](https://user-images.githubusercontent.com/44235769/184449749-1f825f52-21ad-49e4-8e86-6b02b1e0076d.png)

Reply:
I'm sorry, but I don't understand the explanation you've written. You are checking that `cosine` is in the defined range of angles and `cosine` is defined as the scalar product of the direction randomly generated and the direction specified by the user as the "axis" of the lambertian distribution. Therefore, you are selecting the angle between the momentum generated randomly and the "axis" of the lambertian distribution, which, in the black box geometry, is (0, 0, -1), but in another geometry could be any direction. Did I get it wrong? From your comment, it seems that the angle is always relative to the global reference system.

Reply:
Regarding the negative sign, it cannot depend on the geometry, because you are building a generic random lambertian generator, which can be used in any geometry. Given a direction, this generator should provide a momentum which follows a Lambertian distribution around that direction, independently of anything else.